In [20]:
# Instalación de dependencias (necesario para Colab)
try:
    from itertools import combinations
    from itertools import product
    import numpy as np
    import matplotlib.pyplot as plt
    import time
    print("Librerías ya instaladas.")
except ImportError:
    !pip install numpy matplotlib
    import numpy as np
    import matplotlib.pyplot as plt


Librerías ya instaladas.


Parte 1

In [3]:
def calcular_X(resultados):
    """
    resultados: lista o array de longitud 10 con valores 0 o 1
    """
    X = 0
    for i, Mi in enumerate(resultados, start=1):
        X += Mi / (2 ** i)
    return X

In [4]:
resultados = [1, 0, 1, 1, 0, 0, 1, 0, 1, 1]
X = calcular_X(resultados)
print(X)

0.6982421875


In [5]:
# Genera todas las combinaciones posibles de 0 y 1 para 3 lanzamientos
t = product([0, 1], repeat=3) 
print(list(t)) 

[(0, 0, 0), (0, 0, 1), (0, 1, 0), (0, 1, 1), (1, 0, 0), (1, 0, 1), (1, 1, 0), (1, 1, 1)]


In [6]:
# Genera todas las combinaciones posibles de "cara" y "número" para 10 lanzamientos y calculo x
recorrido = set()
for resultados in product([0, 1], repeat=10):
    X = calcular_X(resultados)
    recorrido.add(X)
elementos = len(recorrido)
print(f"Elementos distintos en el recorrido: {elementos}")        

Elementos distintos en el recorrido: 1024


In [7]:

def analizar_X(i):
    recorrido_x = set()
    for resultados in product([0, 1], repeat=i):
        X = calcular_X(resultados)
        recorrido_x.add(X)
    recorrido_x = list(recorrido_x)
    media = np.mean(recorrido_x)
    varianza = np.var(recorrido_x)
    return media, varianza


In [8]:
cantidades = [10, 12, 14, 16, 18, 20]

print(f"{'Monedas':<10} | {'Esperanza':<12} | {'Varianza':<12}")
print("-" * 40)

for k in cantidades:
    E, Var = analizar_X(k)
    print(f"{k:<10} | {E:<12.5f} | {Var:<12.5f}")

Monedas    | Esperanza    | Varianza    
----------------------------------------
10         | 0.49951      | 0.08333     
12         | 0.49988      | 0.08333     
14         | 0.49997      | 0.08333     
16         | 0.49999      | 0.08333     
18         | 0.50000      | 0.08333     
20         | 0.50000      | 0.08333     


Parte 2

In [9]:
def simular_X(monedas):
    resultados = np.random.randint(0, 2, size=monedas)
    return resultados


In [10]:
n = [10**3, 10**4, 10**5, 10**6]
simulaciones = []
def N_simulaciones(veces,monedas):
    simulacion = []
    for i in range(0,veces):
        simulacion.append(calcular_X(simular_X(monedas)))   
    return simulacion
for k in n:
    simulaciones.append(N_simulaciones(k,10))




In [29]:
def frecuencia_Relativa(lista):
    x = 0
    for i in lista:
        if i >= 2/7 and i <= 6/7:
            x += 1
    return x / len(lista)

In [12]:
print(f"{'Simulacion':<10} | {'Frecuencia relativa':<12}")
print("-" * 40)

for k in range(0,4):
    F = frecuencia_Relativa(simulaciones[k])
    print(f"{n[k]:<10} | {F:<12.5} ")

Simulacion | Frecuencia relativa
----------------------------------------
1000       | 0.562        
10000      | 0.5697       
100000     | 0.57004      
1000000    | 0.57129      


In [35]:
cantidades = [10, 20, 30, 40, 50, 60]
n = [10**3, 10**4, 10**5, 10**6]

frecuencias_Relativas = [[], [], [], [], [], []]
tiempos = []
for i in range(0,4):
    inicio = time.perf_counter()
    frecuencia = frecuencia_Relativa(N_simulaciones(n[i], cantidades[0]))
    reinicio1 = time.perf_counter() 
    tiempos.append(reinicio1 - inicio)
    frecuencias_Relativas[0].append(frecuencia)
    frecuencia = frecuencia_Relativa(N_simulaciones(n[i], cantidades[1]))
    reinicio2 = time.perf_counter()
    tiempos.append(reinicio2 - reinicio1)
    frecuencias_Relativas[1].append(frecuencia)
    frecuencia = frecuencia_Relativa(N_simulaciones(n[i], cantidades[2]))
    reinicio3 = time.perf_counter()
    tiempos.append(reinicio3 - reinicio2)
    frecuencias_Relativas[2].append(frecuencia)
    frecuencia = frecuencia_Relativa(N_simulaciones(n[i], cantidades[3]))
    reinicio4 = time.perf_counter()
    tiempos.append(reinicio4 - reinicio3)
    frecuencias_Relativas[3].append(frecuencia)
    frecuencia = frecuencia_Relativa(N_simulaciones(n[i], cantidades[4]))
    reinicio5 = time.perf_counter()
    tiempos.append(reinicio5 - reinicio4)
    frecuencias_Relativas[4].append(frecuencia)
    frecuencia = frecuencia_Relativa(N_simulaciones(n[i], cantidades[5]))
    reinicio6 = time.perf_counter()
    tiempos.append(reinicio6 - reinicio5)
    frecuencias_Relativas[5].append(frecuencia)



for i in range(0,6):
    print(f"{'monedas':<10} | {'Simulacion':<10} | {'Frecuencia relativa':<12} | {'Tiempo (s)':<12}")
    print("-" * 40)
    for j in range(0,4):
        F = frecuencias_Relativas[i][j]
        print(f"{cantidades[i]:<10} | {n[j]:<10} | {F:<12.5f} | {tiempos[i + j*6]:<12.5f} ")

monedas    | Simulacion | Frecuencia relativa | Tiempo (s)  
----------------------------------------
10         | 1000       | 0.59300      | 0.01824      
10         | 10000      | 0.56930      | 0.08718      
10         | 100000     | 0.57051      | 1.35006      
10         | 1000000    | 0.57052      | 8.98080      
monedas    | Simulacion | Frecuencia relativa | Tiempo (s)  
----------------------------------------
20         | 1000       | 0.54900      | 0.02830      
20         | 10000      | 0.56550      | 0.14228      
20         | 100000     | 0.57200      | 1.81486      
20         | 1000000    | 0.57141      | 11.22844     
monedas    | Simulacion | Frecuencia relativa | Tiempo (s)  
----------------------------------------
30         | 1000       | 0.56200      | 0.03516      
30         | 10000      | 0.57650      | 0.13784      
30         | 100000     | 0.57207      | 1.84387      
30         | 1000000    | 0.57162      | 14.28430     
monedas    | Simulacion | Frecuenc

Parte 3

8) Para comprobar que la esperanza y la varianza se aproximan hacia la uniforme continua realizamos 10000 simulaciones para cada cantidad de monedas. Así "quitamos la suerte" del experimento y estabilizamos los valores

In [14]:
def analizar_monedas(k):

    n_simulaciones = 10000
    
    lanzamientos = np.random.randint(0, 2, size=(n_simulaciones, k))
    
    potenciasMi = 2 ** np.arange(1, k + 1)
    
    # Calculamos X para cada una de las 10,000 simulaciones
    # Dividimos cada moneda por su potencia y sumamos la fila
    simulaciones_X = np.sum(lanzamientos / potenciasMi, axis=1)
    
    # Esperanza y varianza de esos resultados
    media = np.mean(simulaciones_X)
    varianza = np.var(simulaciones_X)
    
    return media, varianza

In [15]:

cantidades = [10, 20, 30, 40, 50]

print(f"{'Monedas':<10} | {'Esperanza':<12} | {'Varianza':<12}")
print("-" * 40)

for k in cantidades:
    E, Var = analizar_monedas(k)
    print(f"{k:<10} | {E:<12.5f} | {Var:<12.5f}")


Monedas    | Esperanza    | Varianza    
----------------------------------------
10         | 0.50342      | 0.08217     
20         | 0.50042      | 0.08290     
30         | 0.50656      | 0.08337     
40         | 0.50175      | 0.08387     
50         | 0.49923      | 0.08443     


Cómo se observa en la tabla, a medida que se aumenta la cantidad de monedas, los valores de Esperanza y Varianza se aproximan a 0.5 y 0.0833